### PR TITLE
WG-2: Playercard

### PR DESCRIPTION
**Pull request by DominikHoffendahl**
**FE**
- [x] (FE) Wire all wiki references when clicking on stuff: Player name -> player page, game name -> game page, rank -> leaderboard (Does not exist, but we can point to /ladder/leaderboard in the wiki)
- [x] (FE) Hover over Games to show game infos --> CHANGE TO: make gamename clickable with the wiki reference, as it's mobile and desktop ready today! 
- [x] (FE) Optional: Introduce tabs for years: 2011 | 2012 | 2018 | Weiherlympics 2018, ... (wont do in this feature)
- [x] (FE) Show ELO diff per match +/-
- [x] (FE) Fix player img fit on firefox **Can not reproduce**

**BE**
- [x] (BE) Only provide matches where the player itself attended
- [x] (BE) Order matches by date, latest first
- [x] (BE) Calculate ELO
- [x] (BE) Provide aggregated data for the top section (overall won, lost, per season ofc and off-season). Percent in 0 to 1 please.
- [ ] ~~(BE) Optional: Provide the overall stats for **the last 30 games or last 12 months** (discuss), additionally, so we can see overall winrate of a player but also a recent one, which is the winrate within **the last 30 games or last 12 months**~~
 
image

### CODE REVIEW

File: templates/test.html [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            <img class="gamelist-item-playerimg" src="{{ opponent.img }}" />
+            {% endfor %}
+
+            <span class="gamelist-item-playername">
+                {% for opponent in game.opponents %} {{ opponent.name }}&nbsp; {% endfor %}
+            </span>
+
+            <span class="gamelist-item-score-{{ game.win ? 'win' : 'lose' }}">
+                {{ game.result }}
+            </span>
```
Comment (danielthiel) [inline templates/test.html]: <data> = {
    'player': <playerdata>,
    'games': [<gamedata>],
}

<playerdata> = {
    'elo': <float>,
    'rank': <int>,
    'img': <string>,
    'name': <string>,
    'sumLadderWon': <int>,
    'sumLadderTie': <int>,
    'sumLadderLose': <int>,
    'sumSeasonWon': <int>,
    'sumSeasonTie': <int>,
    'sumSeasonLose': <int>,
    'sumWon': <int>,
    'sumTie': <int>,
    'sumLose': <int>
}


<gamedata> = {
    'opponents': [<playerdata>],
    'result': <string ??>,
    'win': <string ??>
}

Comment (danielthiel) [reply templates/test.html]: wie sollen game.result und game.win aussehen??

Comment (DominikHoffendahl) [reply templates/test.html]: das data object ist aus der sicht des spieler nach meinem template.

man könnte aber auch machen:
```
game = {
    participants: [ UserObject*N ],
    points: [ number*N ],
    winner: UserObject,
    tie: boolean,
}
```

beispiel
```
game = {
    participants: [ UserA, UserB, UserC ],
    points: [15, 10, 1],
    winner: UserA, // oder user_id statt nochmal dem object
    tie: false,
}
```

ka up 2 you :>

Comment (DominikHoffendahl) [reply templates/test.html]:
```
game = {
    participants: [ UserObject*N ],
    points: [ number*N ],
    pointsRule: "highestWins" | "lowestWins",
    winner: UserObject,
    tie: boolean,
}
```

Comment (danielthiel) [reply templates/test.html]: generell: MATCH als name (nicht game), das GAME ist die spielart (man muss ja wissen was gepsielt wurde :D)
```
<matchdata> = {
    'gamename': <string>,
    'participants': [<playerdata>],
    'result': [<integer>], # same order as participants
    'winner': <playerdata>,
    'datetime': <datetime>,
    'tie': <boolean>
}
```


ansonsten noch:
```
OUTPUT SPECIFICATION
<data> = {
    'player': <playerdata>,
    'games': [<gamedata>],
}

<playerdata> = {
    'name': <string>,
    'elo': <float>,
    'rank': <int>,
    'img': <string>,
    'sumLadderWon': <int>,
    'sumLadderTie': <int>,
    'sumLadderLose': <int>,
    'sumSeasonWon': <int>,
    'sumSeasonTie': <int>,
    'sumSeasonLose': <int>,
    'sumWon': <int>,
    'sumTie': <int>,
    'sumLose': <int>
}
```

Comment (danielthiel) [reply templates/test.html]:
```
<matchdata> = {
    'gamename': <string>,
    'participants': [<playerdata>],
    'result': [<integer>], # same order as participants
    'winner': <playerdata>,
    'datetime': <datetime>,
    'tie': <boolean>,
    'pointrule': [<string>] # example: "lowerwins"
}
```

Comment (DominikHoffendahl) [reply templates/test.html]: 1. matchdata fehlt:
- game img
- game id oder game internalName (für wiki reference nacher)

2.
```
<data> = {
    'player': <playerdata>,
    'games': [<gamedata>],
}
```
=
```
<data> = {
    'player': <playerdata>,
    'matches': [<matchdata>],
}
```
?

3.
```
<playerdata> = {
    'name': <string>,
    'elo': <float>,
    'rank': <int>,
    'img': <string>,
    'sumWon': <int>,
    'sumTie': <int>,
    'sumLose': <int>,
    'sumSeasonWon': <int>,
    'sumSeasonTie': <int>,
    'sumSeasonLose': <int>,
    'sumOffSeasonWon': <int>,
    'sumOffSeasonTie': <int>,
    'sumOffSeasonLose': <int>
}
```

Comment (danielthiel) [reply templates/test.html]: OUTPUT SPECIFICATION
```
<data> = {
    'player': <playerdata>,
    'matches': [<matchdata>],
}

<playerdata> = {
    'name': <string>,
    'elo': <float>,
    'rank': <int>,
    'img': <string>,
    'sumOffSeasonWon': <int>,
    'sumOffSeasonTie': <int>,
    'sumOffSeasonLose': <int>,
    'sumSeasonWon': <int>,
    'sumSeasonTie': <int>,
    'sumSeasonLose': <int>,
    'sumWon': <int>,
    'sumTie': <int>,
    'sumLose': <int>,
}


<matchdata> = {
    'game': <gamedata>,
    'participants': [<playerdata>],
    'result': [<integer>], # same order as participants
    'winner': <playerdata>,
    'datetime': <datetime>,
    'tie': <boolean>,
}

<gamedata> = {
    'name': <string>,
    'img': <string,
    'field_of_play': <string>,
    'field_of_play_icon': <string>,
    'pointrules': [<string>] # example: "lowerwins",
    'specific_leveling_value' = <integer>,
}
```

Comment (DominikHoffendahl) [reply templates/test.html]: perfection T_T

File: templates/test.html [hunk 2]
```diff
@@ -208,20 +208,20 @@ <h1>{{ data.player.name }}</h1>
                 </div>
 
                 <div class="gamelist">
-                    {% for game in data.games %}
+                    {% for match in data.matches %}
                     <div class="gamelist-item">
                         <span class="gamelist-item-vs">vs</span>
 
-                        {% for opponent in game.opponents %}
+                        {% for opponent in match.participants %}
```
Comment (DominikHoffendahl) [inline templates/test.html]: gamelist-item braucht noch match.game.name, aktuell weis man nicht welches spiel dort gespielt wurde in der liste

TODO
- design
- template
- data wire

File: scripts/weihergames_data_extractor.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    matches = read_matches_from_json(players, games)
+    process_matches(matches)
+
+    templates_dir = os.path.join(_ROOT_FILE, '../templates')
+    env = Environment(
+        loader=FileSystemLoader(templates_dir),
+        autoescape=select_autoescape(['html', 'xml'])
+    )
+
+    template = env.get_template('test.html')
```
Comment (DominikHoffendahl) [inline scripts/weihergames_data_extractor.py]: please add some utils to the template
https://stackoverflow.com/questions/30104914/how-to-get-current-time-in-jinja

`getYearFromDate(dateString) => number`

Comment (danielthiel) [reply scripts/weihergames_data_extractor.py]: getYearFromDateString its called

Comment (danielthiel) [reply scripts/weihergames_data_extractor.py]: please test

File: scripts/weihergames_data_extractor.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        g.field_of_play = data.get('field_of_play', '')
+        g.field_of_play_icon = data.get('field_of_play_icon', '')
+        g.pointrules = data.get('pointrules', [])
+        g.specific_leveling_value = data.get('specific_leveling_value', 0)
+        games.append(g)
+        logging.debug('Game found in DB: {0}'.format(g))
+    return games
+
+
+def read_matches_from_json(players, games):
```
Comment (DominikHoffendahl) [inline scripts/weihergames_data_extractor.py:289]: reverse the order please, latest matches first

**Comment by danielthiel** [comment]
- [x] please add hover over Games, with the game info (Field of Play, potential pointrules - it is important if lower score is winning to get it known)
- [x] of course also wiki link

**Comment by danielthiel** [comment]
- [x] player picture "zoom" not working in FF (picture not resized), Chrome works

File: scripts/weihergames_data_extractor.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    for name, data in j['members'].items():
+        p = PlayerData(name)
+        p.elo = ELO_START_VALUE
+        p.img = data.get('img', None)
+        players.append(p)
+        logging.debug('Player found in DB: {0}'.format(p))
+    return players
+
+
+def read_games_from_json():
```
Comment (DominikHoffendahl) [inline scripts/weihergames_data_extractor.py:269]: image

Only assign match to data where the player iself attended

Comment (DominikHoffendahl) [reply scripts/weihergames_data_extractor.py:269]: The player (doompy) did not attend crossboccia in this screenshot

File: scripts/weihergames_data_extractor.py [hunk 6]
```diff
@@ -125,9 +133,23 @@ def img(self):
...
         img_path = os.path.join(_IMAGE_FOLDER, str(img_filename))
         if os.path.isfile(img_path):
-            self.__img = get_relative_image_path(img_filename)
+            self.__img = _get_relative_image_path(img_filename)
         else:
-            self.__img = get_relative_image_path(_ALT_PLAYER_IMG)
+            self.__img = _get_relative_image_path(_ALT_PLAYER_IMG)
+
+    def add_match(self, match):
+        if type(match) not MatchData:
```
Comment (DominikHoffendahl) [inline scripts/weihergames_data_extractor.py]: syntax error bei mir :(

image

**Comment by danielthiel** [comment]
grafik
please fix formatting @DominikHoffendahl , prozent reicht zahl ohne komma

**Comment by danielthiel** [comment]
@DominikHoffendahl kannst du im ToolTip noch expected Wert `match.result_expected` der Spieler anzeigen, wird benötigt um auch den ELO calculator gut zu parametrisieren

**Pull request merged**